File: app/src/main/java/com/example/helpwheel/ui/dashboard/DashboardFragment.java
package com.example.helpwheel.ui.dashboard;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.example.helpwheel.R;
import com.example.helpwheel.databinding.FragmentDashboardBinding;
import com.example.helpwheel.utils.Constants;
import com.example.helpwheel.utils.ZoomPageTransformer;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.material.tabs.TabLayoutMediator;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionDeniedResponse;
import com.karumi.dexter.listener.PermissionGrantedResponse;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.single.PermissionListener;
import com.skydoves.balloon.ArrowPositionRules;
import com.skydoves.balloon.Balloon;
import com.skydoves.balloon.BalloonAnimation;
import com.skydoves.balloon.BalloonSizeSpec;

import java.io.IOException;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

public class DashboardFragment extends Fragment {
    private DashboardViewModel dashboardViewModel;
    private FragmentDashboardBinding binding;
    SharedPreferences preferences;
    SharedPreferences.Editor editor;
    private FusedLocationProviderClient mFusedLocationClient;
    String key = "a98ca7720a8fd711bb8548bf2373e263";
    Constants constants;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        dashboardViewModel = new ViewModelProvider(requireActivity()).get(DashboardViewModel.class);
        binding = FragmentDashboardBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        preferences = requireContext().getSharedPreferences(constants.APP_PREFERENCES, Context.MODE_PRIVATE);
        editor = preferences.edit();
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(requireActivity());

        if (preferences.getString(constants.USERNAME, "user").equals("user") || preferences.getString(constants.USERNAME, "user").equals(""))
            showWelcomeScreen();

        changeUi();
        initViewPager();
        initListeners();
        initObservers();
        setupSwitcher();
    }

    private void initObservers() {
        dashboardViewModel.getIsPermissionGranted().observe(getViewLifecycleOwner(), this::updateUiWhenPermissionChanged);
    }

    @Override
    public void onStart() {
        super.onStart();
        requireActivity().findViewById(R.id.customBnb).setVisibility(View.VISIBLE);
        checkPermissions();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void showBalloon() {
        if (preferences.getBoolean(Constants.IS_FIRST_LAUNCHED_DASHBOARD, true)) {
            Balloon balloon = new Balloon.Builder(requireContext())
                    .setWidth(BalloonSizeSpec.WRAP)
                    .setHeight(BalloonSizeSpec.WRAP)
                    .setText(getString(R.string.balloon_edit_profile_here))
                    .setTextColorResource(R.color.white)
                    .setTextSize(15f)
                    .setArrowPositionRules(ArrowPositionRules.ALIGN_ANCHOR)
                    .setArrowSize(10)
                    .setArrowPosition(0.5f)
                    .setPadding(12)
                    .setCornerRadius(8f)
                    .setBackgroundColorResource(R.color.gray)
                    .setBalloonAnimation(BalloonAnimation.ELASTIC)
                    .setLifecycleOwner(getViewLifecycleOwner())
                    .setAutoDismissDuration(5000L)
                    .setMarginRight(15)
                    .build();
            balloon.showAlignBottom(binding.userGreetingEditButton);
        }
    }

    private void updateUiWhenPermissionChanged(Boolean isPermissionGranted) {
        if (isPermissionGranted) {
            binding.enterCityEditText.setVisibility(View.GONE);
            binding.manualInput.setVisibility(View.VISIBLE);
            binding.enterCity.setVisibility(View.GONE);
            binding.weatherBtn.setVisibility(View.GONE);
        } else {
            binding.manualInput.setVisibility(View.GONE);
            binding.enterCityEditText.setVisibility(View.VISIBLE);
            binding.enterCity.setVisibility(View.VISIBLE);
            binding.weatherBtn.setVisibility(View.VISIBLE);
            binding.enterCity.setText((preferences.getString(constants.USER_CITY, getString(R.string.kyiv))));
            dashboardViewModel.getWeatherDataManualInput(preferences.getString(constants.USER_CITY, getString(R.string.kyiv)), key);
        }
    }

    private void initViewPager() {
        ViewPager2 viewPager2 = binding.weatherViewPager;
        FragmentStateAdapter pagerAdapter = new ViewPagerAdapter(requireActivity());
        viewPager2.setAdapter(pagerAdapter);
        viewPager2.setPageTransformer(new ZoomPageTransformer());
        new TabLayoutMediator(binding.tab, binding.weatherViewPager, (tab, position) -> {
        }).attach();
    }

    private void initListeners() {
        binding.userGreetingEditButton.setOnClickListener(view -> showEditDataFragment());
        binding.weatherBtn.setOnClickListener(v -> {
            if (Objects.requireNonNull(binding.enterCity.getText()).toString().trim().equals(""))
                Toast.makeText(binding.getRoot().getContext(), R.string.enter_city_message, Toast.LENGTH_LONG).show();
            else {
                getWeatherFromManualInput();
                editor.putString(constants.USER_CITY, binding.enterCity.getText().toString());
                dashboardViewModel.setCity(binding.enterCity.getText().toString());
            }
        });
    }

    private void setupSwitcher() {
        binding.manualInput.setOnCheckedChangeListener((compoundButton, isChecked) -> {
            if (isChecked) {
                checkPermissions();
            } else {
                binding.enterCity.setText((preferences.getString(constants.USER_CITY, getString(R.string.kyiv))));
                dashboardViewModel.getWeatherDataManualInput(preferences.getString(constants.USER_CITY, getString(R.string.kyiv)), key);
                dashboardViewModel.setIsGpsTurnedOn(true);
            }
            changeInputTypeWeather();
        });
    }

    private void getWeatherFromManualInput() {
        String user_city = Objects.requireNonNull(binding.enterCity.getText()).toString().trim();
        binding.enterCityEditText.setVisibility(View.VISIBLE);
        binding.enterCity.setVisibility(View.VISIBLE);
        binding.weatherBtn.setVisibility(View.VISIBLE);
        dashboardViewModel.getWeatherDataManualInput(user_city, key);
        editor.putString(constants.USER_CITY, user_city).apply();
    }

    private void checkPermissions() {
        if (ActivityCompat.checkSelfPermission(requireContext(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            Dexter.withContext(requireContext())
                    .withPermission(
                            Manifest.permission.ACCESS_FINE_LOCATION
                    ).withListener(new PermissionListener() {
                        @Override
                        public void onPermissionGranted(PermissionGrantedResponse permissionGrantedResponse) {
                            getMyLocation();
                            showBalloon();
                            editor.putBoolean(Constants.IS_FIRST_LAUNCHED_DASHBOARD, false).apply();
                        }

                        @Override
                        public void onPermissionDenied(PermissionDeniedResponse permissionDeniedResponse) {
                            dashboardViewModel.setIsPermissionGranted(false);
                            showBalloon();
                            editor.putBoolean(Constants.IS_FIRST_LAUNCHED_DASHBOARD, false).apply();
                        }

                        @Override
                        public void onPermissionRationaleShouldBeShown(PermissionRequest permissionRequest, PermissionToken permissionToken) {
                            permissionToken.continuePermissionRequest();
                        }
                    }).check();

        } else
            getMyLocation();
    }

    private void showWelcomeScreen() {
        NavHostFragment.findNavController(this).navigate(R.id.action_navigation_dashboard_to_welcomeFragment);
    }

    private void showEditDataFragment() {
        NavHostFragment.findNavController(this).navigate(R.id.action_navigation_dashboard_to_changeDataFragment);
    }

    @SuppressLint("SetTextI18n")
    public void changeUi() {
        binding.greetingText.setText(getString(R.string.hello_user_text) + " " + preferences.getString(constants.USERNAME, "user") + "!");
    }

    @SuppressLint("MissingPermission")
    private void changeInputTypeWeather() {
        if (binding.manualInput.isChecked()) {
            binding.weatherBtn.setVisibility(View.GONE);
            binding.enterCityEditText.setVisibility(View.GONE);
            binding.enterCity.setVisibility(View.GONE);
        } else {
            binding.weatherBtn.setVisibility(View.VISIBLE);
            binding.enterCityEditText.setVisibility(View.VISIBLE);
            binding.enterCity.setVisibility(View.VISIBLE);
        }
    }

    @SuppressLint("MissingPermission")
    private void getMyLocation() {

        mFusedLocationClient.getLastLocation()
                .addOnSuccessListener(requireActivity(), location -> {
                    if (location != null) {
                        if (location.getLatitude() == 0 && location.getLongitude() == 0) {
                            binding.weatherLoading.setVisibility(View.VISIBLE);
                        } else {
                            dashboardViewModel.setIsPermissionGranted(true);
                            dashboardViewModel.setIsGpsTurnedOn(true);
                            dashboardViewModel.getWeatherDataAutomatically(location.getLatitude(), location.getLongitude(), key);
                            Geocoder gcd = new Geocoder(requireContext(), Locale.getDefault());
                            List<Address> addresses = null;
                            try {
                                addresses = gcd.getFromLocation(location.getLatitude(), location.getLongitude(), 1);
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                            if (addresses != null && !addresses.isEmpty()) {
                                String city = addresses.get(0).getLocality();
                                dashboardViewModel.setCity(city);
                            }
                        }

                    } else
                        dashboardViewModel.setIsGpsTurnedOn(false);
                });

    }
}

File: app/src/main/java/com/example/helpwheel/ui/dashboard/DashboardFragment.kt
package com.example.helpwheel.ui.dashboard

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.location.Address
import android.location.Geocoder
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import com.example.helpwheel.R
import com.example.helpwheel.databinding.FragmentDashboardBinding
import com.example.helpwheel.ui.dashboard.viewPagerFragments.DescriptionWeather
import com.example.helpwheel.ui.dashboard.viewPagerFragments.WeatherWindSpeed
import com.example.helpwheel.utils.*
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.material.tabs.TabLayoutMediator
import com.karumi.dexter.Dexter
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.PermissionDeniedResponse
import com.karumi.dexter.listener.PermissionGrantedResponse
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.single.PermissionListener
import java.io.IOException
import java.util.*

class DashboardFragment : Fragment() {
    private lateinit var dashboardViewModel: DashboardViewModel
    private lateinit var binding: FragmentDashboardBinding
    private lateinit var preferences: SharedPreferences
    private lateinit var editor: SharedPreferences.Editor
    private lateinit var mFusedLocationClient: FusedLocationProviderClient

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentDashboardBinding.inflate(inflater, container, false)
        dashboardViewModel = ViewModelProvider(requireActivity())[DashboardViewModel::class.java]
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        preferences = requireContext().getSharedPreferences(APP_PREFERENCES, Context.MODE_PRIVATE)
        editor = preferences.edit()
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(requireActivity())
        if (preferences.getString(USERNAME, "user").equals("user") || preferences.getString(USERNAME, "user").equals(""))
            showWelcomeScreen()
        changeUi()
        initViewPager()
        initListeners()
        initObservers()
        setupSwitcher()
    }


    private fun initObservers() {
        dashboardViewModel.getIsPermissionGranted().observe(viewLifecycleOwner) { updateUiWhenPermissionChanged(it) }
    }

    override fun onStart() {
        super.onStart()
        requireActivity().findViewById<CardView>(R.id.customBnb).visibility = View.VISIBLE
        checkPermissions()
    }


    private fun showBalloon() {
        if (preferences.getBoolean(IS_FIRST_LAUNCHED_DASHBOARD, true))
            BuildBalloon(requireContext(), getString(R.string.balloon_edit_profile_here), viewLifecycleOwner).balloon.showAlignBottom(binding.userGreetingEditButton)
    }

    private fun updateUiWhenPermissionChanged(isPermissionGranted: Boolean) {
        if (isPermissionGranted) {
            binding.enterCityEditText.visibility = View.GONE
            binding.manualInput.visibility = View.VISIBLE
            binding.enterCity.visibility = View.GONE
            binding.weatherBtn.visibility = View.GONE
        } else {
            binding.manualInput.visibility = View.GONE
            binding.enterCityEditText.visibility = View.VISIBLE
            binding.enterCity.visibility = View.VISIBLE
            binding.weatherBtn.visibility = View.VISIBLE
            binding.enterCity.setText((preferences.getString(USER_CITY, getString(R.string.kyiv))))
            preferences.getString(USER_CITY, getString(R.string.kyiv))?.let { dashboardViewModel.getWeatherDataManualInput(it, KEY) }
        }
    }

    private fun initViewPager() {
        val fragmentList = ArrayList<Fragment>()
        fragmentList.add(DescriptionWeather())
        fragmentList.add(WeatherWindSpeed())
        val viewPager2 = binding.weatherViewPager
        val pagerAdapter = ViewPagerAdapter(requireActivity(), fragmentList)
        viewPager2.adapter = pagerAdapter
        viewPager2.setPageTransformer(ZoomPageTransformer())
        TabLayoutMediator(binding.tab, binding.weatherViewPager) { _, _ -> }.attach()
    }

    private fun initListeners() {
        binding.userGreetingEditButton.setOnClickListener { showEditDataFragment() }
        binding.weatherBtn.setOnClickListener {
            if (Objects.requireNonNull(binding.enterCity.text).toString().trim() == "")
                Toast.makeText(binding.root.context, R.string.enter_city_message, Toast.LENGTH_LONG).show()
            else {
                getWeatherFromManualInput()
                editor.putString(USER_CITY, binding.enterCity.text.toString())
                dashboardViewModel.setCity(binding.enterCity.text.toString())
            }
        }
    }

    private fun setupSwitcher() {
        binding.manualInput.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                checkPermissions()
            } else {
                binding.enterCity.setText(
                    (preferences.getString(
                        USER_CITY,
                        getString(R.string.kyiv)
                    ))
                )
                preferences.getString(USER_CITY, requireContext().resources.getString(R.string.kyiv))?.let { dashboardViewModel.getWeatherDataManualInput(it, KEY) }
                dashboardViewModel.setIsGpsTurnedOn(true)
            }
            changeInputTypeWeather()
        }
    }

    private fun getWeatherFromManualInput() {
        val userCity = Objects.requireNonNull(binding.enterCity.text).toString().trim()
        binding.enterCityEditText.visibility = View.VISIBLE
        binding.enterCity.visibility = View.VISIBLE
        binding.weatherBtn.visibility = View.VISIBLE
        dashboardViewModel.getWeatherDataManualInput(userCity, KEY)
        editor.putString(USER_CITY, userCity).apply()
    }

    private fun checkPermissions() {
        if (ActivityCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Dexter.withContext(requireContext())
                .withPermission(
                    Manifest.permission.ACCESS_FINE_LOCATION
                ).withListener(object : PermissionListener {
                    override fun onPermissionGranted(p0: PermissionGrantedResponse?) {
                        getMyLocation()
                        showBalloon()
                        editor.putBoolean(IS_FIRST_LAUNCHED_DASHBOARD, false).apply()
                    }

                    override fun onPermissionDenied(p0: PermissionDeniedResponse?) {
                        dashboardViewModel.setIsPermissionGranted(false)
                        showBalloon()
                        editor.putBoolean(IS_FIRST_LAUNCHED_DASHBOARD, false).apply()
                    }

                    override fun onPermissionRationaleShouldBeShown(
                        p0: PermissionRequest?,
                        p1: PermissionToken?
                    ) {
                        p1!!.continuePermissionRequest()
                    }
                }).check()

        } else
            getMyLocation()
    }

    private fun showWelcomeScreen() {
        NavHostFragment.findNavController(this)
            .navigate(R.id.action_navigation_dashboard_to_welcomeFragment)
    }

    private fun showEditDataFragment() {
        NavHostFragment.findNavController(this).navigate(R.id.action_navigation_dashboard_to_changeDataFragment)
    }

    @SuppressLint("SetTextI18n")
    fun changeUi() {
        binding.greetingText.text = getString(R.string.hello_user_text) + " " + preferences.getString(USERNAME, "user") + "!"
    }

    private fun changeInputTypeWeather() {
        if (binding.manualInput.isChecked) {
            binding.weatherBtn.visibility = View.GONE
            binding.enterCityEditText.visibility = View.GONE
            binding.enterCity.visibility = View.GONE
        } else {
            binding.weatherBtn.visibility = View.VISIBLE
            binding.enterCityEditText.visibility = View.VISIBLE
            binding.enterCity.visibility = View.VISIBLE
        }
    }

    @SuppressLint("MissingPermission")
    private fun getMyLocation() {

        mFusedLocationClient.lastLocation
            .addOnSuccessListener(requireActivity()) { location ->
                if (location != null) {
                    if (location.latitude == 0.0 && location.longitude == 0.0) {
                        binding.weatherLoading.visibility = View.VISIBLE
                    } else {
                        dashboardViewModel.setIsPermissionGranted(true)
                        dashboardViewModel.setIsGpsTurnedOn(true)
                        dashboardViewModel.getWeatherDataAutomatically(
                            location.latitude,
                            location.longitude,
                            KEY
                        )
                        val gcd = Geocoder(requireContext(), Locale.getDefault())
                        var addresses: List<Address>? = null
                        try {
                            addresses =
                                gcd.getFromLocation(
                                    location.latitude,
                                    location.longitude,
                                    1
                                )
                        } catch (e: IOException) {
                            e.stackTrace
                        }
                        if (addresses != null && addresses.isNotEmpty()) {
                            val city = addresses[0].locality
                            dashboardViewModel.setCity(city)
                        }
                    }

                } else
                    dashboardViewModel.setIsGpsTurnedOn(false)
            }
    }
}
